hudi-hadoop-common/src/main/java/org/apache/hudi/common/util/OrcUtils.java

```diff
@@ -117,25 +117,17 @@ public ClosableIterator<HoodieKey> getHoodieKeyIterator(HoodieStorage storage, S
       conf.addResource(HadoopFSUtils.getFs(filePath.toString(), conf).getConf());
       Reader reader = OrcFile.createReader(convertToHadoopPath(filePath), OrcFile.readerOptions(conf));
 
-      Schema readSchema = HoodieAvroUtils.getRecordKeyPartitionPathSchema();
+      Schema readSchema = keyGeneratorOpt
+          .map(keyGenerator -> {
+            List<String> fields = new ArrayList<>();
+            fields.addAll(keyGenerator.getRecordKeyFieldNames());
+            fields.addAll(keyGenerator.getPartitionPathFields());
+            return HoodieAvroUtils.projectSchema(readAvroSchema(storage, filePath), fields);
+          })
+          .orElse(partitionPath.isPresent() ? HoodieAvroUtils.getRecordKeySchema() : HoodieAvroUtils.getRecordKeyPartitionPathSchema());
```

Comment:
present or not present?

Reply:
present, if the field is present in memory then you don't need to read it from the record.